src/Pate/Equivalence.hs

```diff
@@ -470,75 +435,92 @@ memDomPre sym memEqRegion inO inP memEq memPred  = do
-regDomRel sym stO stP regEquiv regDom  = do
-  let
-    regsO = simRegs stO
-    regsP = simRegs stP
-    regRel r p =
-      impM sym (return p) $ applyRegEquivRelation regEquiv r (regsO ^. MM.boundValue r) (regsP ^. MM.boundValue r)
+regDomRel hdr sym stO stP regDom  = do
+  let regRel r p = impM sym (return p) $ registerEquivalence hdr sym stO stP r
   foldr (\(Some r, p) -> andM sym (regRel r p)) (return $ W4.truePred sym) (PER.toList regDom)
 
 eqDomPre ::
   IsSymInterface sym =>
-  MM.RegisterInfo (MM.ArchReg arch) =>
+  PA.ValidArch arch =>
   sym ->
-  -- | stack memory region
-  W4.SymNat sym ->
   SimInput sym arch PBi.Original ->
   SimInput sym arch PBi.Patched ->
-  EquivRelation sym arch ->
+  EquivContext sym arch ->
   PED.EquivalenceDomain sym arch ->
   IO (W4.Pred sym)
-eqDomPre sym stackRegion inO inP eqRel eqDom  = do
+eqDomPre sym inO inP (EquivContext hdr stackRegion) eqDom  = do
   let
     stO = simInState inO
     stP = simInState inP
     
-    regsEq = regDomRel sym stO stP (eqRelRegs eqRel) (PED.eqDomainRegisters eqDom)
+    regsEq = regDomRel hdr sym stO stP (PED.eqDomainRegisters eqDom)
     stacksEq =
-      memDomPre sym (MemRegionEquality $ MT.memEqAtRegion sym stackRegion) inO inP (eqRelStack eqRel) (PED.eqDomainStackMemory eqDom)
+      memDomPre sym (MemEqAtRegion stackRegion) inO inP (PED.eqDomainStackMemory eqDom)
     memEq =
-      memDomPre sym (MemRegionEquality $ MT.memEqOutsideRegion sym stackRegion) inO inP (eqRelMem eqRel) (PED.eqDomainGlobalMemory eqDom)
+      memDomPre sym (MemEqOutsideRegion stackRegion) inO inP (PED.eqDomainGlobalMemory eqDom)
   andM sym regsEq (andM sym stacksEq memEq)
 
 eqDomPost ::
   W4.IsSymExprBuilder sym =>
-  MM.RegisterInfo (MM.ArchReg arch) =>
+  PA.ValidArch arch =>
   sym ->
   SimOutput sym arch PBi.Original ->
   SimOutput sym arch PBi.Patched ->
-  EquivRelation sym arch ->
+  EquivContext sym arch ->
   PED.EquivalenceDomain sym arch ->
   IO (W4.Pred sym)
-eqDomPost sym outO outP eqRel stPred  = do
+eqDomPost sym outO outP (EquivContext hdr stackRegion) stPred = do
   let
     stO = simOutState outO
     stP = simOutState outP
     
-    regsEq = regDomRel sym stO stP (eqRelRegs eqRel) (PED.eqDomainRegisters stPred)
-    stacksEq = memDomPost sym outO outP (eqRelStack eqRel) (PED.eqDomainStackMemory stPred)
-    memEq = memDomPost sym outO outP (eqRelMem eqRel) (PED.eqDomainGlobalMemory stPred)
+    regsEq = regDomRel hdr sym stO stP (PED.eqDomainRegisters stPred)
+    stacksEq = memDomPost sym (MemEqAtRegion stackRegion) outO outP (PED.eqDomainStackMemory stPred)
+    memEq = memDomPost sym (MemEqOutsideRegion stackRegion) outO outP (PED.eqDomainGlobalMemory stPred)
   andM sym regsEq (andM sym stacksEq memEq)
+
+-- | This simply bundles up the necessary state elements necessary to resolve equality.
+data EquivContext sym arch where
+  EquivContext ::
+    { eqCtxHDR :: PA.HasDedicatedRegister arch
+    , eqCtxStackRegion :: W4.SymNat sym
```

Comment:
It's needed just because there are two separate memory domains: one for the stack and one for non-stack memory.

Reply:
It is likely worth noting that this would become more complex if we add support for a more systematic treatment of memory regions. Unless it wouldn't become more complicated (in which case it would be worth noting why)